File: code/replication/referee2_replicate_composition_regression.py
"""
Referee 2 cross-language replication — Python.

Independently reproduces the composition regression from drafts/cfi.qmd.
Pulls FRED series via the public API (no fredr), builds monthly cause-share
dataset, drops the pandemic window (Mar 2020 – Dec 2021), and runs:

    cc_ce ~ temp_layoff + perm_job_loser + job_leaver + reentrant + new_entrant
            + month-of-year FE    (Completed Temp Job omitted)

Also runs ic_ce on the same RHS, computes New Entrant share statistics
(long-run mean, last-12m mean, pre-pandemic mean), and dumps results to a
machine-readable comparison table.

Run:
    python code/replication/referee2_replicate_composition_regression.py
"""
from __future__ import annotations

import io
import json
import os
import sys
from pathlib import Path

import numpy as np
import pandas as pd
import requests
import statsmodels.formula.api as smf

FRED_KEY = "ccd350785a210a7653a20d02ce92d83a"
FRED_URL = "https://api.stlouisfed.org/fred/series/observations"
START = "1994-01-01"
OUT_JSON = Path(__file__).resolve().parent / "referee2_results_python.json"

CAUSE_SERIES = {
    "Temporary Layoff":     "LNS13023653",
    "Permanent Job Losers": "LNS13026638",
    "Completed Temp Job":   "LNS13026637",
    "Job Leavers":          "LNS13023705",
    "Reentrants":           "LNS13023557",
    "New Entrants":         "LNS13023569",
}
TOTAL_SERIES = "UNEMPLOY"
CLAIMS_SERIES = ["CCSA", "ICSA", "COVEMP"]

SHARE_COLS_WITH_PANDEMIC = [
    "temp_layoff_share", "perm_job_loser_share", "completed_temp_share",
    "job_leaver_share", "reentrant_share", "new_entrant_share",
]
RHS_SHARE_COLS = [c for c in SHARE_COLS_WITH_PANDEMIC if c != "completed_temp_share"]


def fred_series(series_id: str, max_retries: int = 5) -> pd.DataFrame:
    import time
    params = {
        "series_id": series_id,
        "api_key": FRED_KEY,
        "file_type": "json",
        "observation_start": START,
    }
    last_err = None
    for attempt in range(max_retries):
        try:
            resp = requests.get(FRED_URL, params=params, timeout=60)
            resp.raise_for_status()
            obs = resp.json()["observations"]
            df = pd.DataFrame(obs)[["date", "value"]]
            df["date"] = pd.to_datetime(df["date"])
            df["value"] = pd.to_numeric(df["value"], errors="coerce")
            df = df.rename(columns={"value": series_id})
            time.sleep(0.4)  # gentle throttle
            return df
        except Exception as e:
            last_err = e
            time.sleep(2 * (attempt + 1))
    raise RuntimeError(f"FRED fetch failed after {max_retries} retries: {last_err}")


def main() -> None:
    # --- Pull monthly cause series + UNEMPLOY
    cause_dfs = []
    for name, sid in CAUSE_SERIES.items():
        d = fred_series(sid).rename(columns={sid: name})
        cause_dfs.append(d)
    cause_wide = cause_dfs[0]
    for d in cause_dfs[1:]:
        cause_wide = cause_wide.merge(d, on="date", how="outer")

    total = fred_series(TOTAL_SERIES).rename(columns={TOTAL_SERIES: "total"})

    # --- Pull weekly claims series; collapse to monthly mean
    claims_frames = []
    for sid in CLAIMS_SERIES:
        d = fred_series(sid)
        d["month"] = d["date"].dt.to_period("M").dt.to_timestamp()
        monthly = d.groupby("month")[sid].mean().reset_index()
        monthly = monthly.rename(columns={"month": "date"})
        claims_frames.append(monthly)
    claims_wide = claims_frames[0]
    for d in claims_frames[1:]:
        claims_wide = claims_wide.merge(d, on="date", how="outer")
    claims_wide["cc_ce"] = claims_wide["CCSA"] / claims_wide["COVEMP"]
    claims_wide["ic_ce"] = claims_wide["ICSA"] / claims_wide["COVEMP"]

    # --- Build regression dataset
    df = cause_wide.merge(total, on="date", how="inner")
    for name in CAUSE_SERIES:
        col = name.lower().replace(" ", "_")
        # The same column naming the QMD uses
        pretty_map = {
            "Temporary Layoff": "temp_layoff_share",
            "Permanent Job Losers": "perm_job_loser_share",
            "Completed Temp Job": "completed_temp_share",
            "Job Leavers": "job_leaver_share",
            "Reentrants": "reentrant_share",
            "New Entrants": "new_entrant_share",
        }
        df[pretty_map[name]] = df[name] / df["total"]
    df["month_of_year"] = df["date"].dt.strftime("%b")
    df["pandemic"] = (df["date"] >= pd.Timestamp("2020-03-01")) & (
        df["date"] <= pd.Timestamp("2021-12-01")
    )
    df = df.merge(claims_wide[["date", "cc_ce", "ic_ce"]], on="date", how="inner")
    reg_df = df[~df["pandemic"]].dropna(
        subset=["cc_ce", "ic_ce"] + RHS_SHARE_COLS
    ).copy()

    formula_rhs = " + ".join(RHS_SHARE_COLS) + " + C(month_of_year)"
    cc_mod = smf.ols(f"cc_ce ~ {formula_rhs}", data=reg_df).fit()
    ic_mod = smf.ols(f"ic_ce ~ {formula_rhs}", data=reg_df).fit()

    def extract(model):
        out = {}
        for col in RHS_SHARE_COLS:
            est = model.params[col]
            se = model.bse[col]
            p = model.pvalues[col]
            out[col] = {
                "estimate_raw": float(est),
                "se_raw": float(se),
                "coef_10pp": float(est * 0.10),
                "se_10pp": float(se * 0.10),
                "p_value": float(p),
            }
        out["_r_squared"] = float(model.rsquared)
        out["_n"] = int(model.nobs)
        return out

    results = {
        "language": "python",
        "package": "statsmodels.OLS",
        "statsmodels_version": __import__("statsmodels").__version__,
        "sample_first_date": str(reg_df["date"].min().date()),
        "sample_last_date": str(reg_df["date"].max().date()),
        "n_obs": int(len(reg_df)),
        "cc_ce": extract(cc_mod),
        "ic_ce": extract(ic_mod),
    }

    # --- New Entrant share descriptive stats (from the share series itself, not the regression sample)
    ne = df[["date", "new_entrant_share"]].dropna().copy()
    ne["pct"] = ne["new_entrant_share"] * 100
    latest_date = ne["date"].max()
    pre_pandemic = ne[ne["date"] <= "2020-02-01"]
    last12 = ne[ne["date"] > (latest_date - pd.DateOffset(months=12))]
    results["new_entrant_share_pct"] = {
        "latest_date": str(latest_date.date()),
        "long_run_mean_1994_present_nonpandemic": float(
            ne.loc[(ne["date"] < "2020-03-01") | (ne["date"] > "2021-12-01"), "pct"].mean()
        ),
        "long_run_mean_1994_present_all": float(ne["pct"].mean()),
        "pre_pandemic_mean_through_2020_02": float(pre_pandemic["pct"].mean()),
        "last_12_months_mean": float(last12["pct"].mean()),
        "latest_value": float(ne.iloc[-1]["pct"]),
    }

    OUT_JSON.write_text(json.dumps(results, indent=2))
    print(json.dumps(results, indent=2))


if __name__ == "__main__":
    main()

File: code/replication/referee2_replicate_composition_regression.R
#!/usr/bin/env Rscript
# Referee 2 cross-language replication — R (standalone, not the QMD).
#
# Independently re-pulls FRED series via fredr, builds the monthly
# cause-share dataset, drops the pandemic window (Mar 2020 – Dec 2021),
# and runs the same OLS specifications as drafts/cfi.qmd to verify
# that the reported numbers are not artifacts of the QMD environment.
#
# Also dumps the long-format panel to CSV so the Stata replication can
# consume an identical input and we isolate OLS-engine discrepancies
# from data-pull discrepancies.
#
# Run:
#   "/c/Program Files/R/R-4.5.3/bin/Rscript.exe" code/replication/referee2_replicate_composition_regression.R

suppressPackageStartupMessages({
  library(tidyverse)
  library(fredr)
  library(jsonlite)
})

fredr_set_key("ccd350785a210a7653a20d02ce92d83a")

here_root <- tryCatch(rprojroot::find_root(rprojroot::is_git_root),
                      error = function(e) getwd())
out_dir <- file.path(here_root, "code", "replication")
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)

cause_series <- c(
  "Temporary Layoff"     = "LNS13023653",
  "Permanent Job Losers" = "LNS13026638",
  "Completed Temp Job"   = "LNS13026637",
  "Job Leavers"          = "LNS13023705",
  "Reentrants"           = "LNS13023557",
  "New Entrants"         = "LNS13023569"
)

cause_monthly <- map_dfr(names(cause_series), \(nm) {
  fredr(series_id = cause_series[[nm]], observation_start = as.Date("1994-01-01")) |>
    transmute(date, cause = nm, value)
}) |>
  pivot_wider(names_from = cause, values_from = value)

total_unemp <- fredr(series_id = "UNEMPLOY", observation_start = as.Date("1994-01-01")) |>
  transmute(date, total = value)

claims_weekly <- bind_rows(
  fredr(series_id = "CCSA",   observation_start = as.Date("1994-01-01")) |>
    transmute(date, series = "CCSA",   value),
  fredr(series_id = "ICSA",   observation_start = as.Date("1994-01-01")) |>
    transmute(date, series = "ICSA",   value),
  fredr(series_id = "COVEMP", observation_start = as.Date("1994-01-01")) |>
    transmute(date, series = "COVEMP", value)
) |>
  mutate(month = floor_date(date, "month")) |>
  group_by(month, series) |>
  summarise(value = mean(value, na.rm = TRUE), .groups = "drop") |>
  pivot_wider(names_from = series, values_from = value) |>
  transmute(
    date  = month,
    cc_ce = CCSA / COVEMP,
    ic_ce = ICSA / COVEMP
  )

reg_df <- cause_monthly |>
  inner_join(total_unemp, by = "date") |>
  mutate(
    temp_layoff_share    = `Temporary Layoff`     / total,
    perm_job_loser_share = `Permanent Job Losers` / total,
    completed_temp_share = `Completed Temp Job`   / total,
    job_leaver_share     = `Job Leavers`          / total,
    reentrant_share      = `Reentrants`           / total,
    new_entrant_share    = `New Entrants`         / total,
    month_of_year        = factor(month.abb[month(date)], levels = month.abb),
    pandemic             = date >= as.Date("2020-03-01") & date <= as.Date("2021-12-01")
  ) |>
  inner_join(claims_weekly, by = "date") |>
  filter(!pandemic) |>
  drop_na(cc_ce, ic_ce, temp_layoff_share, perm_job_loser_share,
          job_leaver_share, reentrant_share, new_entrant_share)

# Dump the merged panel for Stata replication — this is the EXACT same
# dataset the QMD builds, written so Stata can run OLS on identical inputs.
reg_df |>
  select(date, cc_ce, ic_ce,
         temp_layoff_share, perm_job_loser_share, completed_temp_share,
         job_leaver_share, reentrant_share, new_entrant_share,
         month_of_year) |>
  mutate(month_of_year = as.character(month_of_year)) |>
  write_csv(file.path(out_dir, "referee2_regression_panel.csv"))

comp_rhs <- "temp_layoff_share + perm_job_loser_share + job_leaver_share + reentrant_share + new_entrant_share + month_of_year"

cc_mod <- lm(as.formula(paste("cc_ce ~", comp_rhs)), data = reg_df)
ic_mod <- lm(as.formula(paste("ic_ce ~", comp_rhs)), data = reg_df)

extract_rhs <- function(model) {
  k <- c("temp_layoff_share", "perm_job_loser_share", "job_leaver_share",
         "reentrant_share", "new_entrant_share")
  co <- summary(model)$coefficients
  out <- setNames(vector("list", length(k)), k)
  for (nm in k) {
    est <- co[nm, "Estimate"]
    se  <- co[nm, "Std. Error"]
    p   <- co[nm, "Pr(>|t|)"]
    out[[nm]] <- list(
      estimate_raw = unname(est),
      se_raw       = unname(se),
      coef_10pp    = unname(est * 0.10),
      se_10pp      = unname(se * 0.10),
      p_value      = unname(p)
    )
  }
  out$`_r_squared` <- summary(model)$r.squared
  out$`_n`         <- nobs(model)
  out
}

results <- list(
  language = "R",
  package  = "lm() (base R)",
  r_version = paste(R.version$major, R.version$minor, sep = "."),
  sample_first_date = as.character(min(reg_df$date)),
  sample_last_date  = as.character(max(reg_df$date)),
  n_obs = nrow(reg_df),
  cc_ce = extract_rhs(cc_mod),
  ic_ce = extract_rhs(ic_mod)
)

# Descriptive stats on New Entrants share (full merged panel, before the
# regression filter) so we can compare to Python's numbers.
ne_full <- cause_monthly |>
  inner_join(total_unemp, by = "date") |>
  transmute(date, new_entrant_share = `New Entrants` / total * 100)

ne_non_pandemic <- ne_full |> filter(date < as.Date("2020-03-01") | date > as.Date("2021-12-01"))
ne_pre <- ne_full |> filter(date <= as.Date("2020-02-01"))
ne_last12 <- ne_full |> filter(date > max(date) - months(12))

results$new_entrant_share_pct <- list(
  latest_date = as.character(max(ne_full$date)),
  long_run_mean_1994_present_nonpandemic = mean(ne_non_pandemic$new_entrant_share, na.rm = TRUE),
  long_run_mean_1994_present_all         = mean(ne_full$new_entrant_share, na.rm = TRUE),
  pre_pandemic_mean_through_2020_02      = mean(ne_pre$new_entrant_share, na.rm = TRUE),
  last_12_months_mean                    = mean(ne_last12$new_entrant_share, na.rm = TRUE),
  latest_value                           = tail(ne_full$new_entrant_share, 1)
)

writeLines(toJSON(results, pretty = TRUE, auto_unbox = TRUE),
           file.path(out_dir, "referee2_results_R.json"))
print(toJSON(results, pretty = TRUE, auto_unbox = TRUE))
